Improve testing documentation in `CONTRIBUTING.md`

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54904 Improve testing documentation in `CONTRIBUTING.md`**

Differential Revision: [D27407009](https://our.internmc.facebook.com/intern/diff/D27407009)

Commit:
Improve testing documentation in `CONTRIBUTING.md`

[ghstack-poisoned]

Commit:
Update on "Improve testing documentation in `CONTRIBUTING.md`"


Differential Revision: [D27407009](https://our.internmc.facebook.com/intern/diff/D27407009)

[ghstack-poisoned]

samestep commented:
Could you update the table of contents?
```diff
diff --git a/CONTRIBUTING.md b/CONTRIBUTING.md
index 52352cb3b5..5b5c817ccc 100644
--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -7,8 +7,10 @@
   - [Nightly Checkout & Pull](#nightly-checkout--pull)
   - [Codebase structure](#codebase-structure)
   - [Unit testing](#unit-testing)
-    - [Better local unit tests with pytest](#better-local-unit-tests-with-pytest)
+    - [Python Unit Testing](#python-unit-testing)
+    - [Better local unit tests with `pytest`](#better-local-unit-tests-with-pytest)
     - [Running `mypy`](#running-mypy)
+    - [C++ Unit Testing](#c-unit-testing)
   - [Writing documentation](#writing-documentation)
     - [Building documentation](#building-documentation)
       - [Tips](#tips)
```
I should add a lint that does this automatically.

samestep commented on CONTRIBUTING.md:
````diff
@@ -293,42 +293,92 @@ into the repo directory.
 ### Running `mypy`
 
-One of the test suites runs `mypy` on the codebase:
+`mypy` is an optional static type checker for Python. We have a test
+suite that runs `mypy` on the entire codebase:
+
 ```bash
 python test/test_type_hints.py
 ```
+
````
This is actually out of date; while we're updating this section, could we update this as well?
````suggestion
`mypy` is an optional static type checker for Python. We have multiple `mypy`
configs for the PyTorch codebase, so you can run them all using this command:

```bash
for CONFIG in mypy*.ini; do mypy --config="$CONFIG"; done
```

````

samestep commented on CONTRIBUTING.md:
```diff
@@ -293,42 +293,92 @@ into the repo directory.
+
 See [Guide for adding type annotations to
 PyTorch](https://github.com/pytorch/pytorch/wiki/Guide-for-adding-type-annotations-to-PyTorch)
 for more information on how to set up `mypy` and tackle type annotation
 tasks, as well as other ways to run `mypy` besides running that test suite.
```
Similarly here:
```suggestion
for more information on how to set up `mypy` and tackle type annotation tasks.
```

samestep reviewed (changes requested):
Looks great, thanks! Just a couple minor requests

Commit:
Update on "Improve testing documentation in `CONTRIBUTING.md`"


Differential Revision: [D27407009](https://our.internmc.facebook.com/intern/diff/D27407009)

[ghstack-poisoned]

Commit:
Update on "Improve testing documentation in `CONTRIBUTING.md`"


Differential Revision: [D27407009](https://our.internmc.facebook.com/intern/diff/D27407009)

[ghstack-poisoned]

samestep commented on CONTRIBUTING.md:
```diff
@@ -8,8 +8,10 @@
 - [Unit testing](#unit-testing)
-  - [Better local unit tests with pytest](#better-local-unit-tests-with-pytest)
-  - [Running `mypy`](#running-mypy)
+    - [Python Unit Testing](#python-unit-testing)
+    - [Better local unit tests with pytest](#better-local-unit-tests-with-pytest)
+    - [Running `mypy`](#running-mypy)
+    - [C++ Unit Testing](#c-unit-testing)
```
Sorry for the inconvenience; this just needs a couple minor formatting tweaks to fix the failing ["Lint / toc" job](https://github.com/pytorch/pytorch/pull/54904/checks?check_run_id=2242141332):
```suggestion
  - [Python Unit Testing](#python-unit-testing)
  - [Better local unit tests with `pytest`](#better-local-unit-tests-with-pytest)
  - [Running `mypy`](#running-mypy)
  - [C++ Unit Testing](#c-unit-testing)
```

samestep reviewed (changes requested):
Thanks for the update! Just a minor tweak to fix the Table of Contents lint, and then you should be good to go.

Commit:
Update on "Improve testing documentation in `CONTRIBUTING.md`"


Differential Revision: [D27407009](https://our.internmc.facebook.com/intern/diff/D27407009)

[ghstack-poisoned]
